drivers: display: st7789v: Simplify configuration code

caspermeijn opened the pull request:
In preparation of another change I simplified the code for configuration of the st7789v driver. In several steps almost all the data coming from the device tree moved to the new `struct st7789v_config`.  It is const so that it is stored in FLASH.
I also made naming more consistent, made software reset usable and fix compiling with power management.

I tested this with `samples/drivers/display` on `pinetime_devkit0`.

Before:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       25712 B       512 KB      4.90%
            SRAM:       20744 B        64 KB     31.65%
        IDT_LIST:          0 GB         2 KB      0.00%
```

After:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       25756 B       512 KB      4.91%
            SRAM:       20680 B        64 KB     31.56%
        IDT_LIST:          0 GB         2 KB      0.00%
```

caspermeijn committed:
drivers: display: st7789v: Make naming consistent

The names of `struct st7789v_data` varied a bit in the driver code.
Make the name consistently `data`.

Signed-off-by: Casper Meijn <casper@meijn.net>

caspermeijn committed:
drivers: display: st7789v: Move params to `struct config`

Create a new `struct st7789v_config` and fill it with const
configuration parameters needed during initialisation.

Signed-off-by: Casper Meijn <casper@meijn.net>

caspermeijn committed:
drivers: display: st7789v: Move gpio config to `struct config`

Move the gpio config from separate pin/flags/device to
`struct gpio_dt_spec`. This in placed in the const config struct.

Signed-off-by: Casper Meijn <casper@meijn.net>

caspermeijn committed:
drivers: display: st7789v: Move spi config to `struct config`

Move the spi config from separate flags and device to
`struct spi_dt_spec`. This in placed in the const config struct.

Signed-off-by: Casper Meijn <casper@meijn.net>

caspermeijn commented:
I think the failed test is not from my code changes. I will rebase tomorrow and try again.

jfischer-no commented on drivers/display/display_st7789v.c:
```diff
@@ -101,16 +99,11 @@ static void st7789v_exit_sleep(struct st7789v_data *data)
 	gpio_pin_set(data->reset_gpio, ST7789V_RESET_PIN, 0);
 	k_sleep(K_MSEC(20));
-#else
-	st7789v_transmit(p_st7789v, ST7789V_CMD_SW_RESET, NULL, 0);
```
This software reset does not work or what is reasoning?

caspermeijn replied:
@jfischer-no Well, on `main` branch it is currently not possible to enable software reset. To enable software reset you need to leave out the reset pin, however defining a reset pin is required. This indicated to me that nobody needed software reset and that is was never tested. I thought this change would simplify the code.

gmarull replied:
@caspermeijn can you remove the `required` field from the reset-gpios? Then you can obtain the gpio using `GPIO_DT_SPEC_INST_GET_OR(0, reset_gpios, {0})`. This way you can then add both, hard and soft reset options.

caspermeijn replied:
@gmarull I will maintain the status-quo in this PR and open a separate PR for the software reset.

gmarull replied:
I think it's better to address it here (in a separate commit), the only issue with the current status is the `required: True` field in the dt binding.

caspermeijn replied:
Added a commit to fix software reset

jfischer-no commented on drivers/display/display_st7789v.c:
```diff
@@ -79,57 +72,59 @@ static void st7789v_set_lcd_margins(struct st7789v_data *data,
 
-static void st7789v_transmit(struct st7789v_data *data, uint8_t cmd,
-		uint8_t *tx_data, size_t tx_count)
+static void st7789v_transmit(struct st7789v_data *data, struct st7789v_config *config, uint8_t cmd,
```
No, please use `const struct device *dev`
```
static void st7789v_transmit(const struct device *dev, uint8_t cmd,
                             uint8_t *tx_data, size_t tx_count)
```

Same for st7789v_exit_sleep, t7789v_set_mem_area, ...

caspermeijn replied:
I thought this version is more type-safe. I will change to `device *`.

gmarull commented on drivers/display/display_st7789v.c:
```diff
@@ -64,68 +61,63 @@ struct st7789v_data {
-#if DT_INST_NODE_HAS_PROP(0, reset_gpios)
 	k_sleep(K_MSEC(1));
-	gpio_pin_set(data->reset_gpio, ST7789V_RESET_PIN, 1);
+	gpio_pin_set_dt(&config->reset_gpio, 1);
 	k_sleep(K_MSEC(6));
-	gpio_pin_set(data->reset_gpio, ST7789V_RESET_PIN, 0);
+	gpio_pin_set_dt(&config->reset_gpio, 0);
 	k_sleep(K_MSEC(20));
```
```c
if (config->reset_gpio.port != NULL) {
     /* hw reset */
} else {
     /* sw reset */
}
```

gmarull commented on drivers/display/display_st7789v.c:
```diff
@@ -63,69 +60,68 @@ struct st7789v_data {
+static void st7789v_transmit(const struct device *dev, uint8_t cmd,
+			     uint8_t *tx_data, size_t tx_count)
 {
+	struct st7789v_config *config = (struct st7789v_config *)dev->config;
```
```suggestion
	struct st7789v_config *config = dev->config;
```

gmarull commented on drivers/display/display_st7789v.c:
```diff
@@ -63,69 +60,68 @@ struct st7789v_data {
+static void st7789v_set_cmd(const struct device *dev, int is_cmd)
 {
-	gpio_pin_set(data->cmd_data_gpio, ST7789V_CMD_DATA_PIN, is_cmd);
+	struct st7789v_config *config = (struct st7789v_config *)dev->config;
```
```suggestion
	struct st7789v_config *config = dev->config;
```

gmarull commented on drivers/display/display_st7789v.c:
```diff
@@ -63,69 +60,68 @@ struct st7789v_data {
+static void st7789v_set_lcd_margins(const struct device *dev,
+				    uint16_t x_offset, uint16_t y_offset)
 {
+	struct st7789v_data *data = (struct st7789v_data *)dev->data;
```
```suggestion
	struct st7789v_data *data = dev->data;
```

gmarull commented on drivers/display/display_st7789v.c:
```diff
@@ -63,69 +60,68 @@ struct st7789v_data {
-static void st7789v_reset_display(struct st7789v_data *data)
+static void st7789v_reset_display(const struct device *dev)
 {
+	struct st7789v_config *config = (struct st7789v_config *)dev->config;
```
```suggestion
	struct st7789v_config *config = dev->config;
```

gmarull commented on drivers/display/display_st7789v.c:
```diff
@@ -138,21 +134,24 @@ static int st7789v_read(const struct device *dev,
+				 const uint16_t x, const uint16_t y,
+				 const uint16_t w, const uint16_t h)
 {
+	struct st7789v_data *data = (struct st7789v_data *)dev->data;
```
```suggestion
	struct st7789v_data *data = dev->data;
```

gmarull commented on drivers/display/display_st7789v.c:
```diff
@@ -161,7 +160,7 @@ static int st7789v_write(const struct device *dev,
 			 const void *buf)
 {
-	struct st7789v_data *data = (struct st7789v_data *)dev->data;
+	struct st7789v_config *config = (struct st7789v_config *)dev->config;
```
```suggestion
	struct st7789v_config *config = dev->config;
```

caspermeijn replied:
@gmarull After applying the suggestion, the compiler complains about discarding 'const' qualifier. When making all references to `config` const I end up with a last warning for the spi_buf:
```
/home/casper/src/zephyrproject/zephyr/drivers/display/display_st7789v.c: In function 'st7789v_transmit':
/home/casper/src/zephyrproject/zephyr/drivers/display/display_st7789v.c:85:47: warning: initialization discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
   85 |         const struct spi_buf tx_buf = {.buf = &cmd, .len = 1};
      |                                               ^
/home/casper/src/zephyrproject/zephyr/drivers/display/display_st7789v.c:93:47: warning: initialization discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
   93 |         const struct spi_buf tx_buf = {.buf = tx_data, .len = tx_count};
      |                                               ^~~~~~~
```
Any ideas how to fix that?

gmarull replied:
`struct st7789v_config *config` should actually be `const struct st7789v_config *config`

caspermeijn replied:
I added the `const` modifier to the config struct. However I needed to add a `(void *)` to the tx_buf. I am unsure whether that is the best solution.

gmarull commented on drivers/display/display_st7789v.c:
```diff
@@ -223,11 +222,11 @@ static int st7789v_set_contrast(const struct device *dev,
 			      struct display_capabilities *capabilities)
 {
-	struct st7789v_data *data = (struct st7789v_data *)dev->data;
+	struct st7789v_config *config = (struct st7789v_config *)dev->config;
```
```suggestion
	struct st7789v_config *config = dev->config;
```

gmarull commented on drivers/display/display_st7789v.c:
```diff
@@ -263,149 +262,120 @@ static int st7789v_set_orientation(const struct device *dev,
-static void st7789v_lcd_init(struct st7789v_data *p_st7789v)
+static void st7789v_lcd_init(const struct device *dev)
 {
+	struct st7789v_data *data = (struct st7789v_data *)dev->data;
+	struct st7789v_config *config = (struct st7789v_config *)dev->config;
```
```suggestion
	struct st7789v_data *data = dev->data;
	struct st7789v_config *config = dev->config;
```

gmarull commented on drivers/display/display_st7789v.c:
```diff
@@ -263,149 +262,120 @@ static int st7789v_set_orientation(const struct device *dev,
-		return -EPERM;
+	if (!spi_is_ready(&config->spi)) {
+		LOG_ERR("SPI device for LCD is not ready");
+		return -EINVAL;
```
```suggestion
		return -ENODEV;
```

gmarull commented on drivers/display/display_st7789v.c:
```diff
@@ -263,149 +262,120 @@ static int st7789v_set_orientation(const struct device *dev,
-		return -EPERM;
+	if (!device_is_ready(config->reset_gpio.port)) {
+		LOG_ERR("GPIO port for display reset is not ready");
+		return -EINVAL;
```
```suggestion
		return -ENODEV;
```

gmarull commented on drivers/display/display_st7789v.c:
```diff
@@ -263,149 +262,120 @@ static int st7789v_set_orientation(const struct device *dev,
-		return -EPERM;
+	if (!device_is_ready(config->cmd_data_gpio.port)) {
+		LOG_ERR("GPIO port for cmd/DATA port is not ready");
+		return -EINVAL;
```
```suggestion
		return -ENODEV;
```

gmarull commented on drivers/display/display_st7789v.c:
```diff
@@ -263,149 +262,120 @@ static int st7789v_set_orientation(const struct device *dev,
 
 static int st7789v_init(const struct device *dev)
 {
 	struct st7789v_data *data = (struct st7789v_data *)dev->data;
```
```suggestion
```
unused

caspermeijn committed:
drivers: display: st7789v: Fix software reset

Software reset was already implemented in the driver when the
`reset_gpio` is not defined. However it could not be activated
because the `reset_gpio` is a required field in the DTS
binding.

Signed-off-by: Casper Meijn <casper@meijn.net>

caspermeijn committed:
drivers: display: st7789v: Fix power management

Fix building with `CONFIG_PM_DEVICE`. An old PM constant was used.

Signed-off-by: Casper Meijn <casper@meijn.net>

gmarull commented on dts/bindings/display/sitronix,st7789v.yaml:
```diff
@@ -11,7 +11,7 @@ include: spi-device.yaml
     reset-gpios:
       type: phandle-array
-      required: true
+      required: false
```
nit: removing required is equivalent to false

caspermeijn replied:
Personally I like that is was explicitly optional and other properties also have `required: false`. Do you want me to remove all `required: false` from this file?

gmarull reviewed (approved):
Just left a minor comment, but LGTM. Thanks for these improvements!

jfischer-no commented on drivers/display/display_st7789v.c:
```diff
@@ -63,33 +63,36 @@ struct st7789v_data {
 
-	struct spi_buf tx_buf = { .buf = &cmd, .len = 1 };
-	struct spi_buf_set tx_bufs = { .buffers = &tx_buf, .count = 1 };
+	{
```
Why these changes, that bring no benefits at all except more code,  and why are they part of this commit?

caspermeijn replied:
The order of the commit was a mess, because that was easier then rewrite the history. I fixed that now.

`tx_data` needs to be const, because some of that data now comes from `st7789v_config`. 
`tx_buf` needs to be const, because it holds `tx_data`.
The double code is because you can't write to the const `tx_buf`.
I added a new scope, to prevent shadowing the the `tx_buf` name.
I hope this makes sense.

caspermeijn replied:
I don't understand what you mean. 

If you mean that `tx_data` shouldn't be `const`, then the data from `st7789c_config` can't be transmitted directly. This could be solved by copying the data before calling transmit (as it did in the old code).

If you mean something else, could you explain how to solve your concerns?

jfischer-no commented on drivers/display/display_st7789v.c:
```diff
@@ -63,33 +63,36 @@ struct st7789v_data {
 static void st7789v_set_cmd(const struct device *dev, int is_cmd)
 {
-	struct st7789v_config *config = (struct st7789v_config *)dev->config;
+	const struct st7789v_config *config = dev->config;
```
commit https://github.com/zephyrproject-rtos/zephyr/pull/36717/commits/6dbb628b7969fcd74fc011672312c7ce54bff809 introduces st7789v_config, therefore this should be fixed there.

caspermeijn replied:
I squashed these changes and reordered the commits, so that they make more sense.

caspermeijn committed:
drivers: display: st7789v: Remove explicit cast

Remove explicit cast of `data`.

Signed-off-by: Casper Meijn <casper@meijn.net>

caspermeijn committed:
drivers: display: st7789v: Use `device *` for internal functions

The internal functions of this driver needed either `device *`
or `st7789v_data *` as argument. Standardize to
using `device *`.

Signed-off-by: Casper Meijn <casper@meijn.net>

gmarull commented:
@jfischer-no ping

JordanYates reviewed (approved):
SPI changes look fine to me
